### PR TITLE
RW-31 vertical rhythm

### PR DESCRIPTION
**Pull request by left23**
RW-55

This PR attempts to address the vertical rhythm in general by using the `cd-flow` component in place of more component-specific margin and padding rules. Because the `cd-flow` add a margin to the top of elements, I've reversed a lot of the rules where previously there was a top border and top padding and adjustments for the first-child, now there is a bottom border, and adjustments to the last-child.

There are also changes to rules in media queries as I start to check everything across multiple viewport dimensions.

### CODE REVIEW

**Comment by left23** [comment]
@orakili I'm adding you as reviewer but let's not merge this until we can chat.

File: html/themes/custom/common_design_subtheme/components/rw-article/rw-article.css
Comment (orakili) [inline html/themes/custom/common_design_subtheme/components/rw-article/rw-article.css:28]:
```suggestion
  padding: 0 0 var(--cd-flow-space) 0;
```

With this change the padding would have the same value as the margin on the other side of the border. What do you think?

File: html/themes/custom/common_design_subtheme/components/rw-blog/rw-blog.css
Comment (orakili) [inline html/themes/custom/common_design_subtheme/components/rw-blog/rw-blog.css:25]:
```suggestion
  padding: 0 0 var(--cd-flow-space);
```

File: html/themes/custom/common_design_subtheme/components/rw-homepage/rw-homepage.css
Comment (orakili) [inline html/themes/custom/common_design_subtheme/components/rw-homepage/rw-homepage.css:8]: The `1rem` differs from the other places overriding the `--cd-flow-space` which use `px` and actually it doesn't seem to be used at the `.rw-river--disasters .rw-river-article--disaster` below override the margin-top (line 13).

Should this be changed to `16px` and line 13 removed and the padding changed to `padding: 0 0 var(--cd-flow-space)` below?

Also same comment as above about the issue with the `css-vars-ponyfill` not picking up the css var override.

File: html/themes/custom/common_design_subtheme/components/rw-homepage/rw-homepage.css
Comment (orakili) [inline html/themes/custom/common_design_subtheme/components/rw-homepage/rw-homepage.css:13]: This overrides the margin from the `cd-flow`.

**Review by orakili — COMMENTED** [review]
Just a first pass at reviewing the code with some comments related to the use of the `cd-flow`. Some other generic thoughts (not particularly for the RW work):

I find that the use of the `cd-flow` makes the components less self-contained as they somehow need to be aware of where the `cd-flow` class was added and have to override the `--cd-flow-space` variable and the styling can be off if the `cd-flow` class is not added.

Also the `cd-flow` class only deals with the margin top but in all the cases it applies here, there is also a matching padding bottom. A solution is to use  `padding-bottom: var(--cd-flow-space)` and I think that should be added to the `cd-flow` component as a variant like `--cd-flow--top-bottom` (with a better name probably).

Another thought: I'm wondering if rather than adding the `cd-flow` classes to the templates, it would not be better to simply use the `--cd-flow-space` variable directly for the margins and paddings in the components to ensure consistent vertical rhythm. This way the components would be more self-contained but with consistent spacing, similar to using the color and font variables.

Note: I think we should add a section in the [CD README](https://github.com/UN-OCHA/common_design/blob/main/README.md) to explain how to use the `cd-flow` and `cd-bumper` classes if it's meant to be the standard way to control vertical rhythm.

File: html/themes/custom/common_design_subtheme/components/rw-article/rw-article.css
Comment (orakili) [inline html/themes/custom/common_design_subtheme/components/rw-article/rw-article.css:31]: Removing the bottom padding affects sections like the maps and infographics on the country/disaster pages:

maps-cards

File: html/themes/custom/common_design_subtheme/components/rw-article/rw-article.css
Comment (left23) [reply html/themes/custom/common_design_subtheme/components/rw-article/rw-article.css:2]: I think this works. The main place where setting the flow value is useful is for the rivers - so the section title and the following elements `.rw-river__articles` and its children have the same vertical spacing.
In the context where we want a different value, I can define the flow value on a parent element, which as you said will be ignored by the polyfill so in IE11, the value on root will be used.

I am trying the `cd-flow` component here because all the values I mention are already 24px. I don't think variants are needed in this case.

File: html/themes/custom/common_design_subtheme/components/rw-blog/rw-blog.css
Comment (left23) [reply html/themes/custom/common_design_subtheme/components/rw-blog/rw-blog.css:23]: True.

File: html/themes/custom/common_design_subtheme/components/rw-article/rw-article.css
Comment (left23) [reply html/themes/custom/common_design_subtheme/components/rw-article/rw-article.css:28]: I don't think there is a need to mix the function of the `cd-flow` here. Padding is about what happens within a component, the flow is about how the component interacts with other elements around it. I don't mind using variables for padding but I am against using the same variable because they are not the same thing.

File: html/themes/custom/common_design_subtheme/components/rw-homepage/rw-homepage.css
Comment (left23) [reply html/themes/custom/common_design_subtheme/components/rw-homepage/rw-homepage.css:8]: In this case 1rem = 16px which is why I didn't differentiate. I can make the units consistent nearer the end of the project when I replace values with variables.
I've removed the margin declaration so this uses the cd-fow value.

File: html/themes/custom/common_design_subtheme/components/rw-homepage/rw-homepage.css
Comment (left23) [reply html/themes/custom/common_design_subtheme/components/rw-homepage/rw-homepage.css]: I'm not sure if you mean `cd-flow` from the base theme? It appears to be working the way I intended. Am I misunderstanding what you're pointing out?

File: html/themes/custom/common_design_subtheme/components/rw-article/rw-article.css
Comment (left23) [reply html/themes/custom/common_design_subtheme/components/rw-article/rw-article.css:31]: fixed

**Comment by left23** [comment]
@orakili I've responded to the comments, made some changes, and more commits to do with the vertical rhythm and general QA.
This is ready for another review.

File: html/themes/custom/common_design_subtheme/components/rw-article/rw-article.css
Comment (orakili) [reply html/themes/custom/common_design_subtheme/components/rw-article/rw-article.css:28]: Fair enough.

File: html/themes/custom/common_design_subtheme/components/rw-article/rw-article.css
Comment (orakili) [reply html/themes/custom/common_design_subtheme/components/rw-article/rw-article.css:2]: Ok. I would be in a favor of adding a note about that in the common_design_subtheme's README for future maintainers. Making a note for myself for now and I can do that when I do a documentation update of the modules/themes later.

File: html/themes/custom/common_design_subtheme/components/rw-homepage/rw-homepage.css
Comment (orakili) [reply html/themes/custom/common_design_subtheme/components/rw-homepage/rw-homepage.css:8]: Roger that.

File: html/themes/custom/common_design_subtheme/components/rw-homepage/rw-homepage.css
Comment (orakili) [reply html/themes/custom/common_design_subtheme/components/rw-homepage/rw-homepage.css]: II mentioned that already in https://github.com/UN-OCHA/rwint9-site/pull/60#discussion_r689958035 and you changed it. It was just to point out that the `margin-top` from the `cd-flow` class didn't apply because of this `margin` rule.

**Review by orakili — COMMENTED** [review]
Looking good. I just noticed some inconsistency regarding the spacing around the river results (showing X of Y). Hopefully the individual comments below are clear enough.

File: html/modules/custom/reliefweb_rivers/templates/reliefweb-rivers-river.html.twig
Comment (orakili) [inline html/modules/custom/reliefweb_rivers/templates/reliefweb-rivers-river.html.twig:42]:
```suggestion
    {{ results ? results|merge({
```

@left23 Sorry, there was a bug in the https://github.com/UN-OCHA/rwint9-site/pull/67 PR, this suggestion fixes it.

File: html/themes/custom/common_design_subtheme/components/rw-related-content/rw-related-content.css
Comment (left23) [reply html/themes/custom/common_design_subtheme/components/rw-related-content/rw-related-content.css:42]: Ok, good point. I will try to consolidate those rules as they only thing unique is the selectors.

I'm not sure I understand your comment about the variants, or rather, I am not sure how I should achieve this.
For the Blog and Organizations, their River sections has the ID of `river-list` which I think is too generic (as it is also used on the pages with the Advanced search) and doesn't differentiate between the list versus card type layout. Am I missing something?

The Related Content block seems to be missing an ID which might explain why I've used `.rw-document + .rw-river--type-reports .rw-river__articles` instead of something more direct.

In the meantime, I've updated the Organizations page to have the same treatment for the cards as the Blog page.

File: html/themes/custom/common_design_subtheme/components/rw-river/rw-river-results.css
Comment (left23) [reply html/themes/custom/common_design_subtheme/components/rw-river/rw-river-results.css:6]: Done

Comment (orakili) [reply html/themes/custom/common_design_subtheme/components/rw-river/rw-river-results.css]: Looking good, thanks.

File: html/themes/custom/common_design_subtheme/components/rw-related-content/rw-related-content.css
Comment (orakili) [reply html/themes/custom/common_design_subtheme/components/rw-related-content/rw-related-content.css:42]: Hmm, for the blog and organizations, I think we can add something like the following snippet in the `reliefweb-rivers-page.html.twig` template override in the subtheme to add a class to the river list wrapper:

```
{% if river == 'organizations' or river == 'blog' %}
  {% set content = content|merge({
    '#attributes': create_attribute().addClass('rw-river--river-columns'),
  }) %}
{% endif %}
```

Then if needed in the `reliefweb-rivers-river.html.twig` template override we could add a class to the articles if the `rw-river--river-columns` (or whatever class name is chosen) is present in the `attributes`.

For the Appeals and Maps, I guess we could check on the ID.

Same for the Related content actually after adding the`#id` after the line to https://github.com/UN-OCHA/rwint9-site/blob/1ca75688bf052c135d888d0b3b9105a52a2b2f8f/html/modules/custom/reliefweb_entities/src/DocumentTrait.php#L201

What do yo think?

**Review by orakili — COMMENTED** [review]
I checked all the different types of front-end pages and they look good. 

The cd-flow approach seems to be working well. 

Is there anything in particular you'd like me to review in more details?

What else do you think needs to be done regarding the vertical rhythm?

**Comment by left23** [comment]
@orakili If this is good to go, let's merge it. 

The consolidation of Blog, Organizations, Related, Appeals, and Maps isn't forthcoming because the layouts differ across viewports
Eg. Appeals displays as one column at 768px, and the others display as two columns. On desktop, Maps has 3 columns and the others have two columns.
I'll put my attempts in another PR.

**Review by orakili — APPROVED** [review]
Merging.

**Pull request merged**
